language/stdlib/modules/libra_account.mvir

```diff
@@ -57,15 +58,99 @@ module LibraAccount {
         payer: address,
     }
 
+    /// Events
+    // A resource representing the counter used to generate uniqueness under each account. There won't be destructor for
+    // this resource to guarantee the uniqueness of the generated handle.
+    resource EventHandleGenerator {
+        // A monotonically increasing counter
+        counter: u64,
+    }
+
+    // A handle for a event such that:
+    // 1. Other modules can emit event to this handle.
+    // 2. Storage can use this handle to prove the total number of events that happened in the past.
+    resource EventHandle<T: unrestricted> {
+        // Total number of events emitted to this event stream.
+        counter: u64
+        // A globally unique ID for this event stream.
+        guid: bytearray,
+    }
+
+    // Derive a fresh unique id by using sender's EventHandleGenerator. The generated bytearray is indeed unique because it
+    // was derived from the hash(sender's EventHandleGenerator || sender_address). This module guarantees that the
+    // EventHandleGenerator is only going to be monotonically increased and there's no way to revert it or destroy it. Thus
+    // such counter is going to give distinct value for each of the new event stream under each sender. And since we
+    // hash it with the sender's address, the result is guaranteed to be globally unique.
+    fresh_guid(counter: &mut Self.EventHandleGenerator, sender: address): bytearray {
```

Comment:
small nit, can you move these functions to the end of the mvir file? It feels weird for the first thing in `LibraAccount` to be events